algorithms/src/Kokkos_Random.hpp

```diff
@@ -985,6 +985,11 @@ class Random_XorShift64_Pool {
     typename locks_type::HostMirror h_lock =
         Kokkos::create_mirror_view(Kokkos::WithoutInitializing, locks_);
 
+    // if the host mirror is the device view, need to fence here
+    // since the init was async.
+    if (state_.data() == h_state.data())
+      exec.fence("Random_XorShift64_Pool::init UnifiedMemory");
+
```

Comment:
technically you could fence right before the loop (after the generator constructor.